taichi/jit/jit_module.h

```diff
@@ -44,15 +45,53 @@ class JITModule {
     return ret;
   }
 
+  static int get_args_bytes() {
+    return 0;
+  }
+
+  template <typename... Args, typename T>
+  static int get_args_bytes(T t, Args... args) {
+    return get_args_bytes(args...) + sizeof(T);
+  }
+
+  static void init_args_pointers(char *packed_args) {
+    return;
+  }
+
+  template <typename... Args, typename T>
+  static void init_args_pointers(char *packed_args, T t, Args... args) {
+    std::memcpy(packed_args, &t, sizeof(t));
+    init_args_pointers(packed_args + sizeof(t), args...);
+    return;
+  }
+
   // Note: **call** is for serial functions
   // Note: args must pass by value
+  // Note: AMDGPU need to pass args by extra_arg currently
   template <typename... Args>
   void call(const std::string &name, Args... args) {
     if (direct_dispatch()) {
       get_function<Args...>(name)(args...);
     } else {
-      auto arg_pointers = JITModule::get_arg_pointers(args...);
-      call(name, arg_pointers);
+      if (module_arch() == Arch::cuda) {
+#if defined(TI_WITH_CUDA)
```

Comment:
`jit_module.h` should only contain `base class` methods. 

You can modify:
```
virtual void call(const std::string &name,
                    const std::vector<void *> &arg_pointers) {
    TI_NOT_IMPLEMENTED
  }
```

to 

```
virtual void call(const std::string &name,
                    const std::vector<void *> &arg_pointers,
                    const std::vector<uint32_t>& arg_bytes) {
    TI_NOT_IMPLEMENTED
  }
```

and make different implementations in `jit_amd.h` and `jit_cuda.h`

Reply:
I considered this design. If we adopt this design and don't judge module's arch, it will not work well when both enable `TI_WITH_CUDA` and `TI_WITH_AMDGPU`.

Reply:
Basically the idea is that the compiler will create `JITModuleCUDA` if initialized with `ti.init(arch=ti.cuda)`, and `JITModuleAMD` if initialized with `ti.init(arch=ti.amdgpu)`. Can you explain into details in terms of why it doesn't work with both `TI_WITH_CUDA` and `TI_WITH_AMDGPU`?

Reply:
In detail, `TI_WITH_CUDA` and `TI_WITH_AMDGPU` are the control variants in complie-scope. If user enables both `TI_WITH_CUDA` and `TI_WITH_AMDGPU`, we should use somehow to get the arch info. In code. However,  if we use macros(`TI_WITH_CUDA` and `TI_WITH_AMDGPU`) to control the which `call` function we use. 
```
#if defined(TI_WITH_CUDA)
use cuda-related call
#else ifdefined(TI_WITH_AMDGPU)
use amdgpu-related call
#endif
```
We will never call amdgpu-related `call` function, even `ti.init(arch=ti.amdgpu)`